common/deadlock/deadlock.go

```diff
@@ -0,0 +1,222 @@
+		params.ClusterMetadata,
+	}
+	if params.ShardController != nil {
+		roots = append(roots, params.ShardController)
+	}
+	return &deadlockDetector{
+		logger:       params.Logger,
+		healthServer: params.HealthServer,
+		config: config{
+			DumpGoroutines:    params.Collection.GetBoolProperty(dynamicconfig.DeadlockDumpGoroutines, true),
+			FailHealthCheck:   params.Collection.GetBoolProperty(dynamicconfig.DeadlockFailHealthCheck, true),
+			AbortProcess:      params.Collection.GetBoolProperty(dynamicconfig.DeadlockAbortProcess, false),
+			Interval:          params.Collection.GetDurationProperty(dynamicconfig.DeadlockInterval, 30*time.Second),
+			MaxWorkersPerRoot: params.Collection.GetIntProperty(dynamicconfig.DeadlockMaxWorkersPerRoot, 10),
+		},
+		roots: roots,
+	}
+}
+
+func (dd *deadlockDetector) Start() error {
+	for _, root := range dd.roots {
+		loopCtx := &loopContext{
+			dd:   dd,
+			root: root,
+			ch:   make(chan common.PingCheck),
+		}
+		dd.loops.Go(loopCtx.run)
+	}
+	return nil
+}
+
+func (dd *deadlockDetector) Stop() error {
+	dd.loops.Cancel()
+	// don't wait for workers to exit, they may be blocked
+	return nil
+}
+
+func (dd *deadlockDetector) detected(name string) {
+	dd.logger.Error("deadlock detected", tag.Name(name))
+
+	if dd.config.FailHealthCheck() {
+		dd.logger.Info("marking grpc services unhealthy")
+		dd.healthServer.Shutdown()
+	}
+
+	if dd.config.DumpGoroutines() {
+		dd.dumpGoroutines()
+	}
+
+	if dd.config.AbortProcess() {
+		dd.logger.Fatal("deadlock detected", tag.Name(name))
+	}
+}
+
+func (dd *deadlockDetector) dumpGoroutines() {
+	profile := pprof.Lookup("goroutine")
+	if profile == nil {
+		dd.logger.Error("could not find goroutine profile")
+		return
+	}
+	var b strings.Builder
+	err := profile.WriteTo(&b, 1) // 1 is magic value that means "text format"
+	if err != nil {
+		dd.logger.Error("failed to get goroutine profile", tag.Error(err))
+		return
+	}
+	// write it as a single log line with embedded newlines.
+	// the value starts with "goroutine profile: total ...\n" so it should be clear
+	dd.logger.Info(b.String())
+}
+
+func (lc *loopContext) run(ctx context.Context) error {
+	lc.ctx = ctx
+	// start with one worker
+	go lc.worker()
+	atomic.StoreInt32(&lc.workers, 1)
+
+	for {
+		// ping blocks until it has passed all checks to a worker goroutine (using an
+		// unbuffered channel).
+		lc.ping([]common.Pingable{lc.root})
+
+		select {
+		case <-time.After(lc.dd.config.Interval()):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (lc *loopContext) ping(pingables []common.Pingable) {
+	for _, pingable := range pingables {
+		for _, check := range pingable.GetPingChecks() {
+			select {
+			case lc.ch <- check:
+			case <-lc.ctx.Done():
+			default:
+				// maybe add another worker if blocked
+				w := atomic.LoadInt32(&lc.workers)
+				if w < int32(lc.dd.config.MaxWorkersPerRoot()) && atomic.CompareAndSwapInt32(&lc.workers, w, w+1) {
+					go lc.worker()
```

Comment:
once started, do we ever reduce worker?

Reply:
no. the reason I wanted to do it on demand is that namespace registry and cluster metadata only need 1-2 workers, but to lock all history shard locks we should use more. I figured scaling on demand was nicer than static configuration.